### PR TITLE
Update Experience Event class for Datastream Config Override Feature

### PR DESCRIPTION
**Pull request by addb**
## Description

Add datastream config override using sendEvent API:
```
let expEventWithConfigOverride = ExperienceEvent(xdm: ["test": "value"], datastreamIdOverride: "DATA_STREAM_ID", datastreamConfigOverride: ["testConfig": "testValue"])

Edge.sendEvent(experienceEvent: expEventWithConfigOverride)
```

Event Payload:
```
[AEP SDK TRACE - <EventHub>] Dispatching Event #Optional(15) -     [
      id: 8856BF19-E3A0-4581-A27F-3F4A848DA874
      name: AEP Request Event
      type: com.adobe.eventType.edge
      source: com.adobe.eventSource.requestContent
      data: {
  "xdm" : {
    "test" : "value"
  },
  "config" : {
    "datastreamIdOverride" : "DATA_STREAM_ID",
    "datastreamConfigOverride" : {
      "testConfig" : "testValue"
    }
  }
}
```

### Changes:
- Modified Experience Event constructor and added convenience constructors for optional fields.
- Added new fields for datastreamIdOverride and datastreamConfigOverride 
- Added unit tests for the change

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/ExperienceEvent.swift [hunk 1]
```diff
@@ -24,31 +24,59 @@ public class ExperienceEvent: NSObject {
...
-    @objc public let datasetIdentifier: String?
+    @objc public private(set) var datasetIdentifier: String?
 
     /// Initialize an Experience Event with the provided event data
     /// - Parameters:
     ///   - xdm:  XDM formatted data for this event, passed as a raw XDM Schema data dictionary.
     ///   - data: Any free form data in a [String : Any] dictionary structure.
     ///   - datasetIdentifier: The Experience Platform dataset identifier where this event should be sent to; if not provided, the default dataset identifier set in the Edge configuration is used
-    @objc public init(xdm: [String: Any], data: [String: Any]? = nil, datasetIdentifier: String? = nil) {
+    @objc public convenience init(xdm: [String: Any], data: [String: Any]? = nil, datasetIdentifier: String? = nil) {
```
[the remark this answers is not in the document]
Comment (addb) [reply Sources/ExperienceEvent.swift:41]: Yes, this is intentional and we dont want the user to use all the three together, and also it will help in deprecating the datasetIdentifier which has to be migrated to datastreamConfigOverride route.

File: Sources/ExperienceEvent.swift [hunk 2]
```diff
@@ -24,31 +24,59 @@ public class ExperienceEvent: NSObject {
     /// Optional free-form query data associated with this event
     @objc public var query: [String: Any]?
 
+    /// Adobe Experience Platform datastream identifier, used to override the default datastream identifier set in the Edge Configuration
+    @objc public private(set) var datastreamIdOverride: String?
+
+    /// Adobe Experience Platform datastream configuration, used to override the datastream configuration
```
Comment (emdobrin) [inline Sources/ExperienceEvent.swift:30]: Nit doc (same for all occurrences):
Datastream identifier used to override the default datastream identifier set in the Edge configuration for this event
Datastream configuration used to override individual settings from the default datastream configuration for this event

File: Sources/ExperienceEvent.swift [hunk 3]
```diff
@@ -24,31 +24,59 @@ public class ExperienceEvent: NSObject {
...
         self.xdm = xdm
         self.data = data
-        self.datasetIdentifier = datasetIdentifier
     }
 
     /// Initialize an Experience Event with the provided event data
     /// - Parameters:
     ///   - xdm: XDM formatted event data passed as an XDMSchema
     ///   - data: Any free form data in a [String : Any] dictionary structure.
-    public init(xdm: XDMSchema, data: [String: Any]? = nil) {
```
Comment (emdobrin) [inline Sources/ExperienceEvent.swift:45]: can we keep this one separate too

File: Tests/UnitTests/ExperienceEventTests.swift [hunk 4]
```diff
@@ -172,4 +180,65 @@ class ExperienceEventTests: XCTestCase {
...
+        }
+        XCTAssertTrue(NSDictionary(dictionary: actualEventData).isEqual(to: expectedEventData))
+    }
+
+    func testAsDictionary_withXdmAndDatastreamIdOverrideAndDatastreamConfigOverride() {
+
+        let expectedXdm = generateXdm()
+        var expectedConfig = generateConfig()
+
+        let expectedDatastreamId = expectedConfig["datastreamIdOverride"] as? String ?? ""
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply Tests/UnitTests/ExperienceEventTests.swift:229]: +1 for improved readability and clear intent, and I would add generateConfig does not seem to save much code and now you have to look into two places to see the expectation data. I think you can create the expectedConfig per test to be more clear for the reader

File: Tests/UnitTests/ExperienceEventTests.swift [hunk 5]
```diff
@@ -172,4 +180,65 @@ class ExperienceEventTests: XCTestCase {
...
+        let expectedDatastreamConfig = expectedConfig["datastreamConfigOverride"] as? [String: Any] ?? [:]
+
+        var expectedEventData: [String: Any] = [:]
+        expectedEventData[xdm] = expectedXdm
+        expectedEventData[config] = expectedConfig
+
+        let experienceEvent = ExperienceEvent(xdm: expectedXdm, datastreamIdOverride: expectedDatastreamId, datastreamConfigOverride: expectedDatastreamConfig)
```
Comment (emdobrin) [inline Tests/UnitTests/ExperienceEventTests.swift:236]: can you add some tests for the other init with XDMSchema?

File: Sources/ExperienceEvent.swift (hunk 2 above)
Comment (emdobrin) [reply Sources/ExperienceEvent.swift]: Can we use "Datastream configuration .. " for the property description too so that it does not get confused with a setting in Experience Platform

File: Tests/UnitTests/ExperienceEventTests.swift [hunk 6]
```diff
@@ -132,6 +125,60 @@ class ExperienceEventTests: XCTestCase {
...
+
+        let experienceEvent = ExperienceEvent(xdm: generatedXdmSchema, data: expectedData, datastreamIdOverride: "testDatastreamId", datastreamConfigOverride: ["key": "value"])
+        guard let actualEventData = experienceEvent.asDictionary() else {
+            XCTFail("Failed to retrieve experience event asDictionary")
+            return
+        }
+        XCTAssertTrue(NSDictionary(dictionary: actualEventData).isEqual(to: expectedEventData))
+    }
+
+    func testAsDictionary_withXdmSchemaAndDataAndDatastreamIdOverrideAndDatastreamConfigOverride() {
```
Comment (emdobrin) [inline Tests/UnitTests/ExperienceEventTests.swift:164]: these two test names L146 and L164 seem reversed for their logic 😛

File: Tests/FunctionalTests/NoConfigFunctionalTests.swift [hunk 7]
```diff
@@ -36,11 +36,11 @@ class NoConfigFunctionalTests: TestBase {
         resetTestExpectations()
         mockNetworkService.reset()
     }
-
```
Comment (emdobrin) [inline Tests/FunctionalTests/NoConfigFunctionalTests.swift:39]: @addb do you know why these lines were not caught by swift lint with the previous prs? 
and secondly, can we rather fix these formatting issues in a separate PR in dev?

File: Tests/UnitTests/ExperienceEventTests.swift [hunk 8]
```diff
@@ -124,6 +125,60 @@ class ExperienceEventTests: XCTestCase {
...
+
+        let experienceEvent = ExperienceEvent(xdm: generatedXdmSchema, data: expectedData, datastreamIdOverride: "testDatastreamId")
+        guard let actualEventData = experienceEvent.asDictionary() else {
+            XCTFail("Failed to retrieve experience event asDictionary")
+            return
+        }
+        XCTAssertTrue(NSDictionary(dictionary: actualEventData).isEqual(to: expectedEventData))
+    }
+
+    func testAsDictionary_withXdmSchemaAndDataAndDatastreamConfigOverride() {
```
Comment (emdobrin) [inline Tests/UnitTests/ExperienceEventTests.swift:146]: can we also test with XdmSchema with empty dataset which should be the expected flow if you use datastream overrides

File: Tests/FunctionalTests/NoConfigFunctionalTests.swift (hunk 7 above)
Comment (addb) [reply Tests/FunctionalTests/NoConfigFunctionalTests.swift:39]: So these are just format warnings so it will not fail the `make lint` command. 

I will open a pr in dev branch as well.

**Review by emdobrin — APPROVED** [review]
